L1Trigger/Phase2L1ParticleFlow/src/newfirmware/egamma/L1EGPuppiIsoAlgo.cc

```diff
@@ -0,0 +1,138 @@
+#include "L1Trigger/Phase2L1ParticleFlow/src/newfirmware/egamma/L1EGPuppiIsoAlgo.h"
+
+using namespace l1ct;
+
+L1EGPuppiIsoAlgo::L1EGPuppiIsoAlgo(const edm::ParameterSet& pSet)
+    : config_(pSet.getParameter<std::string>("pfIsoType"),
+              pSet.getParameter<double>("pfPtMin"),
+              pSet.getParameter<double>("dZMax"),
+              pSet.getParameter<double>("dRMin"),
+              pSet.getParameter<double>("dRMax"),
+              pSet.getParameter<bool>("pfCandReuse")) {}
+
+void L1EGPuppiIsoAlgo::run(const EGIsoObjsEmu& l1EGs,
+                           const PuppiObjs& l1PFCands,
+                           EGIsoObjsEmu& outL1EGs,
+                           z0_t z0) const {
+  outL1EGs.reserve(l1EGs.size());
+
+  // make a list of pointers to PF candidates
+  // the pointer will be removed from the list once the candidate has been used
+  std::list<std::shared_ptr<PuppiObj>> workPFCands;
+  std::list<std::shared_ptr<PuppiObj>> workPFCandsPV;
+  for (unsigned i = 0; i < l1PFCands.size(); ++i) {
+    workPFCands.emplace_back(std::make_shared<PuppiObj>(l1PFCands[i]));
+    workPFCandsPV.emplace_back(std::make_shared<PuppiObj>(l1PFCands[i]));
+  }
+
+  for (const auto& l1EG : l1EGs) {
+    auto outL1EG(l1EG);
+    iso_t iso = 0;
+    iso_t isoPV = 0;
+    if (!workPFCands.empty()) {
+      iso = calcIso(l1EG, workPFCands);
+      isoPV = calcIso(l1EG, workPFCandsPV, z0);
+    }
+
+    if (config_.pfIsoType_ == L1EGPuppiIsoAlgoConfig::kPFIso) {
+      outL1EG.setHwIso(EGIsoObjEmu::IsoType::PfIso, iso);
+      outL1EG.setHwIso(EGIsoObjEmu::IsoType::PfIsoPV, isoPV);
+    } else {
+      outL1EG.setHwIso(EGIsoObjEmu::IsoType::PuppiIso, iso);
+      outL1EG.setHwIso(EGIsoObjEmu::IsoType::PuppiIsoPV, isoPV);
+    }
+    outL1EGs.emplace_back(outL1EG);
+  }
+}
+
+void L1EGPuppiIsoAlgo::run(EGIsoObjsEmu& l1EGs, const PuppiObjs& l1PFCands, z0_t z0) const {
+  // make a list of pointers to PF candidates
+  // the pointer will be removed from the list once the candidate has been used
+  std::list<std::shared_ptr<PuppiObj>> workPFCands;
+  std::list<std::shared_ptr<PuppiObj>> workPFCandsPV;
+  for (unsigned i = 0; i < l1PFCands.size(); ++i) {
+    workPFCands.emplace_back(std::make_shared<PuppiObj>(l1PFCands[i]));
+    workPFCandsPV.emplace_back(std::make_shared<PuppiObj>(l1PFCands[i]));
+  }
+
+  for (auto& l1EG : l1EGs) {
+    iso_t iso = 0;
+    iso_t isoPV = 0;
+    if (!workPFCands.empty()) {
+      iso = calcIso(l1EG, workPFCands);
+      isoPV = calcIso(l1EG, workPFCandsPV, z0);
+    }
+
+    if (config_.pfIsoType_ == L1EGPuppiIsoAlgoConfig::kPFIso) {
```

Comment:
It is actually an enum and should be rather lightweight to evaluate.
The possibility to use PF candidates instead of PUPPI ones stems from an earlier, more general version of the algorithm. This could be simplified if it has to only handle PUPPI now.

Reply:
Yes, my comment referred to the previous version where the parameter was a string.